common/achievement/create.ts

```diff
@@ -0,0 +1,96 @@
+    });
+    return userAchievement;
+}
+
+async function getOrCreateUserAchievement<ID extends ActionID>(template: Achievement_template, userId: string, context?: SpecificNotificationContext<ID>) {
+    const isGlobal =
+        template.templateFor === achievement_template_for_enum.Global ||
+        template.templateFor === achievement_template_for_enum.Global_Courses ||
+        template.templateFor === achievement_template_for_enum.Global_Matches;
+    const existingUserAchievement = await findUserAchievement(template.id, userId, !isGlobal && context);
+    if (!existingUserAchievement) {
+        return await createAchievement(template, userId, context);
+    }
+    return existingUserAchievement;
+}
+
+async function createAchievement<ID extends ActionID>(currentTemplate: Achievement_template, userId: string, context: SpecificNotificationContext<ID>) {
+    const templatesByGroup = await getAchievementTemplates(TemplateSelectEnum.BY_GROUP);
+    const keys = Object.keys(context);
+    const userAchievementsByGroup = await prisma.user_achievement.findMany({
+        where: {
+            template: {
+                group: currentTemplate.group,
+            },
+            userId,
+            AND: keys.map((key) => {
+                return {
+                    context: {
+                        path: key,
+                        equals: context[key],
+                    },
+                };
+            }),
+        },
+        orderBy: { template: { groupOrder: 'asc' } },
+    });
+
+    const nextStepIndex = userAchievementsByGroup.length > 0 ? userAchievementsByGroup.findIndex((e) => e.groupOrder === currentTemplate.groupOrder) + 1 : 0;
+
+    const templatesForGroup = templatesByGroup.get(currentTemplate.group);
+    if (templatesForGroup && templatesForGroup.length > nextStepIndex) {
+        const createdUserAchievement = await createNextUserAchievement(templatesForGroup, nextStepIndex, userId, context);
+        return createdUserAchievement;
+    }
+}
+
+async function createNextUserAchievement<ID extends ActionID>(
+    templatesForGroup: Achievement_template[],
+    nextStepIndex: number,
+    userId: string,
+    context: SpecificNotificationContext<ID>
+) {
+    const nextStepTemplate = templatesForGroup[nextStepIndex];
+    // Here a user template is created for the next template in the group. This is done to always have the data availible for the next step.
+    // This could mean to, for example, have the name of a match partner that is not yet availible due to a unfinished matching process.
+    if (nextStepTemplate && nextStepTemplate.isActive) {
+        const createdUserAchievement = await prisma.user_achievement.create({
+            data: {
+                userId: userId,
+                group: nextStepTemplate.group,
+                groupOrder: nextStepTemplate.groupOrder,
+                context: context ? context : Prisma.JsonNull,
```

Comment:
I think we should call some transformation function here. Just to make sure that the context is always of the same format

Reply:
context has the type SpecificNotificationContext and therefore needs no transformation (right format). Prisma.JsonNull is used to pass an empty object to prisma db because prisma can't handle {}. The prisma documentation recomends the usage of Prisma.JsonNull.

Reply:
Can you guarantee that there is no way to pass other data than `SpecificNotificationContext`?